lib/Dialect/Calyx/CalyxOps.cpp

```diff
@@ -182,6 +182,28 @@ LogicalResult calyx::verifyControlLikeOp(Operation *op) {
   }
   return verifyControlBody(op);
 }
+Operation *calyx::getExternHWModule(OpBuilder &builder, ComponentOp op) {
+  SmallVector<hw::PortInfo, 8> hwPortInfos;
+  auto addHWPortInfo = [&](auto portInfos) {
+    for (auto portInfo : enumerate(portInfos)) {
+      hw::PortInfo hwPortInfo;
+      bool isInput = portInfo.value().direction == calyx::Direction::Input;
+      hwPortInfo.direction =
+          isInput ? hw::PortDirection::INPUT : hw::PortDirection::OUTPUT;
+      hwPortInfo.type = portInfo.value().type;
+      hwPortInfo.argNum = portInfo.index();
+      hwPortInfo.name = portInfo.value().name;
+      hwPortInfos.push_back(hwPortInfo);
+    }
+  };
+
+  addHWPortInfo(op.getInputPortInfo());
+  addHWPortInfo(op.getOutputPortInfo());
+
+  auto extOp = builder.create<hw::HWModuleExternOp>(
+      op->getLoc(), builder.getStringAttr(op.getName()), hwPortInfos);
+  return extOp;
```

Comment:
Any reason why you're not just returning the value instead of creating an `extOp`? It was probably undergo RVO, but just curious.

Reply:
No specific reason, will change.